dCommon/AMFDeserialize.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "BitStream.h"
+
+#include <vector>
+#include <string>
+
+class AMFValue;
+class AMFDeserialize {
+	public:
+		/**
+		 * Read an AMF3 value from a bitstream.
+		 * 
+		 * @param inStream inStream to read value from.
+		 * @return Returns an AMFValue with all the information from the bitStream in it.
+		 */
+		static AMFValue* Read(RakNet::BitStream* inStream, bool clearKeys = true);
+	private:
+		/**
+		 * @brief Private method to read a U29 integer from a bitstream
+		 * 
+		 * @param inStream bitstream to read data from
+		 * @return The number as an unsigned 29 bit integer
+		 */
+		static uint32_t ReadU29(RakNet::BitStream* inStream);
+
+		/**
+		 * @brief Reads a string from a bitstream
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return The read string
+		 */
+		static std::string ReadString(RakNet::BitStream* inStream);
+		
+		/**
+		 * @brief Read an AMFDouble value from a bitStream
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return Double value represented as an AMFValue
+		 */
+		static AMFValue* ReadAmfDouble(RakNet::BitStream* inStream);
+
+		/**
+		 * @brief Read an AMFArray from a bitStream
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return Array value represented as an AMFValue
+		 */
+		static AMFValue* ReadAmfArray(RakNet::BitStream* inStream);
+
+		/**
+		 * @brief Read an AMFString from a bitStream
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return String value represented as an AMFValue
+		 */
+		static AMFValue* ReadAmfString(RakNet::BitStream* inStream);
+
+		/**
+		 * @brief Read an AMFInteger from a bitStream
+		 * 
+		 * @param inStream bitStream to read data from
+		 * @return Integer value represented as an AMFValue
+		 */
+		static AMFValue* ReadAmfInteger(RakNet::BitStream* inStream);
+
+		/**
+		 * List of strings read so far saved to be read by reference.
+		 */
+		static std::vector<std::string> accessedElements;
```

Comment:
I made this static because of your comment earlier in the thread lol.  I'll make this not static anymore.

Reply:
Right, you mean https://github.com/DarkflameUniverse/DarkflameServer/pull/599#discussion_r906762914m, I presume. Sorry about that, your original design there was fine in that respect. I guess my emphasis was meant to be more on *method on `AMFArrayValue`* (which has to be static because you don't have an instance yet), rather than on *consider making this [..] static*. Should have caught that earlier.